keps/sig-auth/3331-structured-config-for-oidc-authentication/README.md

```diff
@@ -0,0 +1,638 @@
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+This enhancement proposal covers implementing the structured configuration for the OIDC authenticator.
+OIDC authentication is essential part of Kubernetes, yet it has limitations in its current state. 
+Bellow we will discuss that limitation and propose solutions.
+
+# Motivation
+
+Structured config for OIDC authentication: noted in various contexts over the past few years. We want to migrate
+away from a flag-based config that is growing without bounds to a proper versioned config format. This would allow us to
+better support various features that have been requested.
+
+### Goals
+
+There are features users want to tune. We need to provide customization of the following:
+
+- Claims validation rules: current OIDC provider supports only audience claim validation and only by exact values
+- Claim mappings: it is only possible to pick a single value from a single claim and prefix groups
+- Use more than one OIDC provider: the only option, for now, is to use [Dex](https://dexidp.io/) (or similar software)
+  as a lightweight OIDC proxy to connect many providers to Kubernetes
+- Change authenticator settings without restarting kube-apiserver
+
+### Non-Goals
+
+- Monitoring
+
+## Proposal
+
+1. Add new `authentication` API object to parse a structured config file `OIDCConfiguration`.
+2. Add a single flag for kube-apiserver to point to the structured config file, automatically reload it on changes.
+3. Use an expression language to let users write their own logic for mappings and validation rules 
+  (expressions should be simple for common cases, yet powerful to cover most user stories).
+
+### Risks and Mitigations
+
+Since this is a new optional feature, no migration is required. Before the Stable release of the feature,
+we should provide examples of migrating from a flag-based config to a new structured config.
+
+## Design Details
+
+### Configuration file
+
+The main part of this proposal is a configuration file. It contains an array of providers:
+
+```go
+type OIDCConfiguration struct {
+    metav1.TypeMeta
+    // Providers is a list of OIDC providers to authenticate Kubernetes users.
+    Providers []Provider
+}
+```
+
+Each provider has several properties that will be described in detail below.
+
+```go
+type Provider struct {
```

Comment:
when possible, prefer naming the file fields to match the flags (dropping the --oidc prefix, translating hyphens into camelCase, etc)

for example:

* clientID
* issuerURL
* groupsClaim
* groupsPrefix

it should be trivial for someone using the flags to translate their current config into a file-based config in order to start using additional aspects of the file-based config (without requiring translating them into CEL expressions and invoking entirely new code paths using CEL)

Reply:
Ok, I will adjust the spec.